### ISSUE TITLE
SEC-18 RLP decoder unsafe allocation

### ISSUE THREAD

**Issue by Gustav-Simonsson** [ethereum/go-ethereum#420]
An attacker can cause the Go client to allocate an arbitrary amount of memory by setting a length prefix in a byte string to a large value.

The RLP decoding function at https://github.com/ethereum/go-ethereum/blob/develop/rlp/decode.go#L28  looks only at the length prefix which allows for up to 256^8 sized strings; larger than available memory in computers today.

Fix: Add maximum size even for types which formally are unbounded. Modify the RLP decoder to reject length prefixes of larger value than can be needed for any byte array: https://github.com/ethereum/wiki/wiki/RLP#definition

**Comment by fjl** [comment]
I have thought about this for a bit.

The decoder **does** validate the size of all values against the size of an enclosing RLP list. This happens in `Stream.Kind`. It would be sufficient to only check the size of the top-level value.

Most RLP is decoded from p2p messages. It would be possible to check that the top level size is not bigger than the declared size of a message there. This can be done with a single call to `Stream.Kind` before decoding the value.

We can also check the top level size in package rlp when decoding from a byte slice, which would cover most other uses.

**Comment by zelig** [comment]
:+1:

**Comment by fjl** [comment]
For additional validation, I will also add the ability to declare a maximum size in a struct tag. 

```go
type foo struct {
    Something []byte `rlp:"maxsize:64k"`
}
```

**Comment by fjl** [comment]
This is not fixed yet.

**Comment by fjl** [comment]
Fixed by #711.
